statvar_imports/us_crash/fars_crashdata/download_script.py

```diff
@@ -0,0 +1,111 @@
+import download_config
+_SCRIPT_PATH = os.path.dirname(os.path.abspath(__file__))
+GCS_OUTPUT_DIR = os.path.join(_SCRIPT_PATH, 'gcs_output')
+if not os.path.exists(GCS_OUTPUT_DIR):
+            os.makedirs(GCS_OUTPUT_DIR)
+sys.path.append(os.path.join(_SCRIPT_PATH, '../../../util/')) 
+from download_util_script import _retry_method
+
+
+def download_and_extract_data(url, input_folder, year):
+    """Downloads zip, extracts ACCIDENT.CSV (handling nested dirs), with retry"""
+    try:
+        response = _retry_method(url, None, 3, 5, 2)
+        filename = os.path.basename(url)
+        input_folder = os.path.join(GCS_OUTPUT_DIR, input_folder)
+        filepath = os.path.join(input_folder, filename)
+
+        if not os.path.exists(input_folder):
+            os.makedirs(input_folder)
+            logging.info(f"Created directory: {input_folder}")
+
+        with open(filepath, 'wb') as f:
+            for chunk in response.iter_content(chunk_size=8192):
+                f.write(chunk)
+        logging.info(f"Downloaded: {filename} to {input_folder}")
+
+        if filename.endswith(".zip"):
+            try:
+                with zipfile.ZipFile(filepath, 'r') as zip_ref:
+                    if zip_ref.namelist(): 
+                        found_accident_csv = False
+                        for item in zip_ref.infolist():
+                            if item.filename.upper().endswith("ACCIDENT.CSV"):
+                                parts = item.filename.split("/")
+                                if len(parts) > 1: 
+                                    output_filename = f"ACCIDENT_{year}.csv"
+                                    output_filepath = os.path.join(input_folder, output_filename)
+                                else:  
+                                    output_filename = f"ACCIDENT_{year}.csv"
+                                    output_filepath = os.path.join(input_folder, output_filename)
+
+                                with zip_ref.open(item) as source, open(output_filepath, "wb") as target:
+                                    shutil.copyfileobj(source, target)
+                                logging.info(f"Extracted {output_filename} from {filename}")
+                                found_accident_csv = True
+
+                        if not found_accident_csv:
+                            logging.info(f"ACCIDENT.CSV not found in {filename}")
+
+                    else:
+                        logging.info(f"Zip file {filename} is empty.")
+
+
+            except Exception as e:
+                logging.fatal(f"An error occurred during extraction from {filename}: {e}")
+                raise RuntimeError(
+                    'Failed to extract files from zip. Check logs for details.') from e
+    except requests.exceptions.HTTPError as e:
+
+        if e.response.status_code == 404:
+            return "no_data_at_url" 
+
+
+def main(argv):
+
+    year = download_config.start_year 
+    while True:
```

Comment:
Does the break from loop rely solely on the HTTPError from download_and_extract_data ? Can we add additional checks based on year as well to avoid infinite loops?

Reply:
Added as discussed.. for year check not going over the current year.